cdap-proto/src/main/java/co/cask/cdap/proto/ProtoConstraint.java

```diff
@@ -0,0 +1,264 @@
+        && getClass() == o.getClass()
+        && getMillisAfterTrigger() == ((DelayConstraint) o).getMillisAfterTrigger();
+    }
+
+    @Override
+    public int hashCode() {
+      return (int) (getMillisAfterTrigger() ^ (getMillisAfterTrigger() >>> 32));
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Delay(%d ms)", getMillisAfterTrigger());
+    }
+  }
+
+  /**
+   * Represents a last run constraint in REST requests/responses.
+   */
+  public static class LastRunConstraint extends ProtoConstraint {
+
+    protected final long millisSinceLastRun;
+
+    public LastRunConstraint(long millisSinceLastRun) {
+      super(Type.LAST_RUN);
+      this.millisSinceLastRun = millisSinceLastRun;
+      validate();
+    }
+
+    public long getMillisSinceLastRun() {
+      return millisSinceLastRun;
+    }
+
+    @Override
+    public void validate() {
+      validateInRange(millisSinceLastRun, "millisSinceLastRun", 1L, null);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      return this == o || o != null
+        && getClass() == o.getClass()
+        && getMillisSinceLastRun() == ((LastRunConstraint) o).getMillisSinceLastRun();
+    }
+
+    @Override
+    public int hashCode() {
+      return (int) (getMillisSinceLastRun() ^ (getMillisSinceLastRun() >>> 32));
+    }
+
+    @Override
+    public String toString() {
+      return String.format("LastRun(%d ms)", getMillisSinceLastRun());
+    }
+  }
+
+  /**
+   * Represents a time range constraint in REST requests/responses.
+   */
+  public static class TimeRangeConstraint extends ProtoConstraint {
+
+    private static final Pattern SIMPLE_TIME_PATTERN = Pattern.compile("^[0-9][0-9]:[0-9][0-9]$");
+
+    protected final String timeZone;
+    protected final String startTime;
+    protected final String endTime;
+
+    public TimeRangeConstraint(String startTime,
+                               String endTime,
+                               TimeZone timeZone) {
+      super(Type.TIME_RANGE);
+      this.startTime = startTime;
+      this.endTime = endTime;
+      this.timeZone = timeZone.getID();
+      validate();
+    }
+
+    public String getStartTime() {
+      return startTime;
+    }
+
+    public String getEndTime() {
+      return endTime;
+    }
+
+    public String getTimeZone() {
+      return timeZone;
+    }
+
+    @Override
+    public void validate() {
+      validateSimpleTime("start", startTime);
+      validateSimpleTime("end ", endTime);
+      ProtoConstraint.validateNotNull(timeZone, "time zone");
+      TimeZone.getTimeZone(timeZone);
+    }
+
+    private void validateSimpleTime(String name, String time) {
+      if (time == null) {
+        throw new IllegalArgumentException(name + " time must not be null");
+      }
+      if (!SIMPLE_TIME_PATTERN.matcher(time).matches()) {
+        throw new IllegalArgumentException(
+          String.format("%s time must be of the form 'hh:mm' but is '%s'", name, time));
```

Comment:
I can change that. For a non-techie user, however, the mixed case would be rather confusing. They don't know that HH is 24h format and hh is 12h format. But ok, it is more accurate